File: 260.single-number-iii.c
int *singleNumber(int *nums, int numsSize, int *returnSize)
{
    int xor = 0;
    for (int i = 0; i < numsSize; ++i)
        xor ^= nums[i];
    int n = 0;
    while ((xor&1) == 0)
    {
        n++;
        xor >>= 1;
    }
    int a = 0, b = 0;
    for (int i = 0; i < numsSize; ++i)
    {
        if ((nums[i] >> n & 1) == 0)
            a ^= nums[i];
        else
            b ^= nums[i];
    }
    int *res = malloc(sizeof(int) * 2);
    *returnSize = 2;
    res[0] = a;
    res[1] = b;
    return res;
}

File: 260.single-number-iii.go
func singleNumber(nums []int) []int {
    xorAll := xor(nums)
    xorFlag := 1
    for xorAll != 0 && xorAll % 2 == 0 {
        xorFlag <<= 1
        xorAll >>= 1
    }
    r1, r2 := 0, 0
    for i := range nums {
        if nums[i] | xorFlag == nums[i] {
            r1 ^= nums[i]
        } else {
            r2 ^= nums[i]
        }
    }
    return []int{r1, r2}
}

func xor(nums []int) int {
    res := 0
    for i := range nums {
        res ^= nums[i]
    }
    return res
}
